cloudant-sync-datastore-android/src/main/java/com/cloudant/sync/replication/ReplicationService.java

```diff
@@ -0,0 +1,319 @@
+package com.cloudant.sync.replication;
+
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+import android.net.wifi.WifiManager;
+import android.os.Binder;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.IBinder;
+import android.os.Looper;
+import android.os.Message;
+import android.support.v4.content.WakefulBroadcastReceiver;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This abstract class forms the basis for creating replication policies on Android.
+ * The replications run in a {@link Service} so that they can properly manage the
+ * lifecycle and handle being killed or restarted by the operating system
+ */
+public abstract class ReplicationService extends Service
+        implements ReplicationPolicyManager.ReplicationsCompletedListener {
+
+    public static final String EXTRA_INTENT = "intent";
+    public static final String EXTRA_COMMAND = "command";
+    public static final int COMMAND_NONE = -1;
+    public static final int COMMAND_START_REPLICATION = 0;
+    public static final int COMMAND_STOP_REPLICATION = 1;
+
+    private Handler mServiceHandler;
+    private ReplicationPolicyManager mReplicationPolicyManager;
+
+    private final Set<ReplicationCompleteListener> mListeners = new HashSet<ReplicationCompleteListener>();
```

Comment:
You should add a note here that all modifications or iterations over `mListeners` should be protected by a synchronized block on the `mListeners` object.